docs(ssr): add experimental app directory guide (#4568)
TanStack/query: docs/guides/ssr.md

- Static Generation (SSG)
- Server-side Rendering (SSR)

> Only Server-side Rendering (SSR) is presently supported by React Server Components in Next.js 13's experimental `app` directory.

React Query supports both of these forms of pre-rendering regardless of what platform you may be using

### Using `initialData`

> The setup is minimal and this can be a quick solution for some cases, but there are a **few tradeoffs to consider** when compared to the full approach:

Reviewer: Love this section! It's been sorely missing, glad you included it, and well written. :smile:

  - Query refetching is based on when the page loads instead of when the data was prefetched on the server
- Prefetch the query on the server, dehydrate the cache and rehydrate it on the client with `<Hydrate>`
  - Requires slightly more setup up front
  - No need to prop drill
  - Query refetching is based on when the query was prefetched on the server

### `<QueryClientProvider>` is required by both the `initialData` and `<Hydrate>` prefetching approaches

The hooks provided by the `react-query` package need to retrieve a `QueryClient` from their context. Wrap your component tree with `<QueryClientProvider>` and pass it an instance of `QueryClient`.

```tsx
// app/providers.jsx
'use client'

import { QueryClient, QueryClientProvider } from '@tanstack/react-query'

export default function Providers({children}) {
  const [queryClient] = React.useState(() => new QueryClient())
  
  return (
    <QueryClientProvider client={queryClient}>
      {children}
    </QueryClientProvider>
  )
}
```

```tsx
// app/layout.jsx
import Providers from './providers'

export default function RootLayout({children}) {
  return (
    <html>

>       <head></head>

Reviewer 2: Good catch 🕺

>       <head></head>

Reviewer 3: wow, big TIL, @efilion. Thanks for that!

It still looks to me that even when that change lands, it will be a workaround based on specific use-cases. If it's not necessary, I'd still suggest to omit it.

You make a good point about `create-next-app` though. I'm not aware of **AppDir** as part of the CLI template though. I'd agree 100% in keeping it consistent with the default template if there's already one.

      <body>
        <Providers>{children}</Providers>
      </body>
    </html>
  )
}
```

### Using `initialData`

Fetch your initial data in a Server Component higher up in the component tree, and pass it to your Client Component as a prop.

```tsx
// app/page.jsx
export default async function Home() {
  const initialData = await getPosts()
  
  return <Posts posts={initialData} />
}
```

```tsx
// app/posts.jsx
'use client'

import { useQuery } from '@tanstack/react-query'

export function Posts(props) {
  const {data} = useQuery({
    queryKey: ['posts'],
    queryFn: getPosts,
    initialData: props.posts,
  })
  
  // ...
}
```

### Using `<Hydrate>`

Create a request-scoped singleton instance of `QueryClient`. **This ensures that data is not shared between different users and requests, while still only creating the QueryClient once per request.**

```tsx
// app/getQueryClient.jsx
import { QueryClient } from '@tanstack/query-core'
import { cache } from 'react'

const getQueryClient = cache(() => new QueryClient())
export default getQueryClient
```

> Re-export the `<Hydrate>` component as a Client Component. This allows `<Hydrate>` to execute on the client and make use of client-only features. **TODO: Plans to add 'use client' to react-query modules?**

Reviewer 4: yes, we have plans to do this. @Ephem this is something we could even do for v4, right?

  - Retrieve the `QueryClient` singleton instance
  - Prefetch the data using the client's prefetchQuery method and wait for it to complete
  - Use `dehydrate` to obtain the dehydrated state of the prefetched queries from the query cache
  - Wrap the component tree that needs the prefetched queries inside your `<Hydrate>` Client Component, and provide it with the dehydrated state

> NOTE: TypeScript currently complains of a type error when using async Server Components. As a temporary workaround, use `{/* @ts-expect-error Server Component */}` when calling this component inside another. For more information, see [End-to-End Type Safety](https://beta.nextjs.org/docs/configuring/typescript#end-to-end-type-safety) in the Next.js 13 beta docs.

```tsx
// app/hydratedPosts.jsx
import { dehydrate } from '@tanstack/query-core'
import getQueryClient from './getQueryClient'
import HydrateOnClient from './hydrateOnClient'

export default async function HydratedPosts() {
  const queryClient = getQueryClient()
  await queryClient.prefetchQuery(['posts'], getPosts)
  const dehydratedState = dehydrate(queryClient, {
    dehydrateMutations: false,

>     shouldDehydrateQuery: query => true // TODO: matchQuery or extend dehydrate

Reviewer 4: agree, we should probably just use the default settings here. There likely are no mutations, so setting `dehydrateMutations` to `false` doesn't accomplish much

As demonstrated, it's fine to prefetch some queries and let others fetch on the queryClient. This means you can control what content server renders or not by adding or removing `prefetchQuery` for a specific query.

> #### With experimental `app` directory

Reviewer 5: I'd add a big `:::caution` under here that this is experimental and is subject to change awaiting _{{some link to a React-issue or RFC}}_

Wrap the `<QueryClientProvier>` and `<Hydrate>` components inside a Client Component that initializes the `QueryClient` your app will use on the client.

- `QueryClient` should be created on an instance ref (or in React State ). **This ensures that data is not shared between different users and requests, while still only creating the QueryClient once per component lifecycle.**
- The `dehydratedState` prop will contain your prefetched data along with `dataUpdatedAt` which provides the time the data was fetched on the server

```tsx
// app/react-query.jsx
'use client'

import { QueryClient, QueryClientProvider, Hydrate } from '@tanstack/react-query'

export default function ReactQuery({dehydratedState, children}) {
  const [queryClient] = React.useState(() => new QueryClient())
  
  return (
    <QueryClientProvider client={queryClient}>

>       <Hydrate state={dehydratedState}>

Reviewer 4: I think we should be able to decouple `QueryClientProvider` and `Hydrate`. The Provider needs to go at the top, once. Nothing changes here really.

`Hydrate` can appear multiple times in the app tree. It just takes data and puts it into the cache if it's newer, as opposed to `initialData`, which will only write data once to the cache if there is no data yet.

Author: Decoupling it is a really good point. In the current implementation, you might end up prefetching data at the top level for a leaf component that is conditionally rendered. Effectively, you could end up prefetching data that isn't needed in the end. Fetching data closer to the component that needs it could also allow you to parameterize that fetch with arguments that wouldn't be available at the top level.

I'll play around with these ideas a bit and see what kind of new implementation I can get working. Thinking something along the lines of a higher-order component that does the fetching and wraps its argument in a `Hydrate`.

Reviewer: I'd say keep it simple for now, the API is probably going to change in the next RQ major anyway. If you wrap `QueryClientProvider` higher up in the tree, you can just do:

```jsx
'use client';
import { Hydrate } from '@tanstack/react-query'

export default function HydrateOnClient(dehydratedState) {
  return <Hydrate state={dehydratedState}>{children}</Hydrate>;
}
```

and use that directly in the Server Components. Don't think there is need for a HoC or something more complex here (at least yet).

Author: https://codesandbox.io/p/github/efilion/next13-react-query/decoupled-hydrate?file=%2Fapp%2Fpage.tsx

The HoC isn't strictly necessary, but it helps hide some of the complexity and keep things DRY.

To take advantage of the deduplication provided by QueryClient, the instance defined at the page level will need to be threaded down through to all components looking to prefetch data. That nullifies one of the advantages of the Hydration approach over the InitialData approach. It would be nice to have something akin to an IoC container that could provide a QueryClient as a request-scoped singleton, but I didn't see much in the way of options to accomplish that at the moment. InversifyJS and TSyringe don't fit the bill here. Likely a difficulty with the fact that being single-threaded, Node.js doesn't provide each request its own thread. Nest.js has a solution, so maybe Next.js could implement this itself down the line. Alternatively, can look at the Reader Monad pattern from functional programming to handle dependency injection.

Reviewer: Yeah, that's a tricky one, clever solution!

I worry about putting that much complexity in the docs though, especially for an experimental approach, might fit better as an example to link to? There might be another way as well, not saying it's better but I thought I'd throw it out there.

So behind the scenes we have RSCs, client (server rendered) and client (browser rendered). For both of the client parts, it's important to have the same `queryClient` throughout the whole rendering which is why we want to hoist it to the top, but for the RSCs, your initial approach is actually fine where you create a new `queryClient` for each RSC (fine as in works, not as in pretty, DRY, fast etc). The reason for this is that since that data gets de/rehydrated anyway, we only need the `queryClient` as a way to be able to call `prefetchQuery` and get the right format back when we call `dehydrate`. The fetches themselves will be deduped behind the scenes anyway, so even if two RSCs call the same endpoint, it at least wont double fetch.

So for RSCs specifically, it's fine to duplicate the `queryClient` instead of passing it around. Ugly and not something we want long term, but might be easier and more concise to document, although possibly more confusing?

Author: > > but it helps hide some of the complexity and keep things DRY.
> 
> I guess another way of stating the above is that I'm not sure this is something we should aim to do at this point, I kind of like having the full extent of the experimental and raw nature of this fully visible, it becomes less of an "officially endorsed pattern" that way which is something that might bite us later (even with caveats, some devs will just copy-paste it). sweat_smile
> 
> Maybe I worry too much though.

I see what you mean. My HoC approach could be a better match for a blog post.

Author: https://codesandbox.io/p/github/efilion/next13-react-query/hydrated-content?file=%2Fapp%2FhydratedContent.tsx

Mainly looking to improve on line 12 inside `hydratedContent.tsx`. Not setting anything at all here might be the best approach for the moment. Might need to make a change to `dehydrate` to pick and choose which queries to dehydrate based on queryKey. Open to suggestions.

        {children}
      </Hydrate>
    </QueryClientProvider>
  )
}
```

Prefetch your data inside a Server Component and provide your previously defined Client Component with the dehydrated state of the server-side `QueryClient`.

> - Create a new QueryClient instance **for each page request. This ensures that data is not shared between users and requests.**

Reviewer 5: I'd rephrase this somehow

```suggestion
- Create a new QueryClient instance **within the body of the React component**. This ensures that data is not shared between users and requests.
```

- Prefetch the data using the clients prefetchQuery method and wait for it to complete
- Use dehydrate to dehydrate the query cache and pass it to your wrapped client provider via the dehydratedState prop
- Wrap your component tree with this new Server Component provider

```tsx
// app/providers.jsx
import { QueryClient, dehydrate } from '@tanstack/query-core'
import ReactQuery from './react-query'

export default async function Providers({children}) {
  const queryClient = new QueryClient()

>   await queryClient.prefetchQuery(['posts'], getPosts)

Reviewer 5: Seems really backwards that you gotta `await queryClient.prefetchQuery(['posts'], getPosts)` in order for the state to be hydrated... 

I would expect some magic to hydrate anything that has been called with `useQuery(..., { suspense: true })` automatically

Reviewer 5: Is there a missing API in React for this at the moment? No way of pushing state from a server-render React-component to a client-side one without this stuff?

Author: Server-side suspense is a whole other wrinkle I hadn't considered. Dropping https://github.com/reactwg/react-18/discussions/37 here for reference because I'll likely want to circle back on this.

Reviewer 4: I agree that this could be handled better by TanStack/query, but for now, it is what it is :). This is exactly why I'd like the whole section to be marked as `experimental` - because the recommendations might change.

also looping in @Ephem :)

Reviewer: I mean, the recommendation should always be to prefetch things as early as possible, both to get a headstart and to avoid waterfalls where you suspend in A to be able to render B which then suspend again etc. We should definitely investigate supporting the other case as well though (if it's even going to be possible), but that's for the future.

One caveat is that currently you _have_ to await the data in the RSC, which blocks streaming what you can early. What we really want is to _start_ prefetching in the RSC but not await it, instead suspending in the component that needs this data. That will require both that React implements de/hydration of Promises and a bunch of changes to RQ though, so I think this PR (with the proposed changes elsewhere) documents the best current experimental approach.

Reviewer: Oh, and like we talked about on Discord @KATT, I think it's interesting here to think about how to support "normal" streaming rendering without RSCs as well. A lot of frameworks might implement streaming rendering without RSCs, so that's also something we want to support.

Reviewer: Btw, for now we should definitely document that `useQuery(..., { suspense: true })` wont work on the server unless you have already prefetched that data.

Author: If you've already prefetched your data, is there any utility to set a client-side suspense boundary? If you don't prefetch, you can set suspense boundaries inside a client-side component and use `useQuery(..., { suspense: true })`. Data will start fetching after the page loads. If you're prefetching, you can set a suspense boundary in a server-side component, and there's no need for `{ suspense: true }` inside `useQuery`. Data will start fetching as soon as possible, the server will send a response without waiting for it, and then stream the rendered component to be integrated into the page once the fetching completes.

Create a request-scoped singleton instance of `QueryClient`. **This ensures that data is not shared between different users and requests, while still only creating the QueryClient once per request.**

> > NOTE: Use the `query-core` module when importing TanStack dependencies within server-side modules (e.g. in getQueryClient.jsx and the hydratedPosts.jsx example Server Component). The `react-query` module only works within Client Components.

Reviewer 6: @efilion is there a V3 version of this ? `query-core` is a V4 module ?

Reviewer 4: no. we're writing the docs for v4 here :)

Reviewer 6: No worries, it made sense for me to just update to V4
